tools/inc/stljobs.h

```diff
@@ -0,0 +1,618 @@
+        assert(io != nullptr);
+        WaitForThreadpoolIoCallbacks(io, TRUE);
+        CloseThreadpoolIo(io);
+        io = nullptr;
+        return std::move(fileHandle);
+    }
+
+    void wait(const bool cancelPending) noexcept {
+        WaitForThreadpoolIoCallbacks(io, cancelPending);
+    }
+
+private:
+    handle<invalid_handle_value_policy> fileHandle{};
+    PTP_IO io{};
+};
+
+struct output_collecting_pipe {
+    static constexpr unsigned long kernelBufferSize = 4096;
+    static constexpr unsigned long bufferSize       = 4096;
+
+    output_collecting_pipe() {
+        // generate a random name for the pipe (we must use a named pipe because anonymous pipes from CreatePipe can't
+        // be used in asynchronous mode)
+        std::random_device rd;
+        constexpr size_t pipeNameBufferCount        = 15 + 8 * 8 + 1;
+        //                                                123456789012345
+        wchar_t pipeNameBuffer[pipeNameBufferCount] = LR"(\\.\pipe\Local\)";
+        wchar_t* pipeNameCursor                     = pipeNameBuffer + 15;
+        for (int values = 0; values < 8; ++values) {
+            unsigned int randomValue = rd();
+            for (int hexits = 0; hexits < 8; ++hexits) {
+                *pipeNameCursor++ = L"0123456789ABCDEF"[randomValue & 0xFu];
+                randomValue >>= 4;
+            }
+        }
+
+        *pipeNameCursor = L'\0';
+
+        auto readHandle = create_named_pipe(pipeNameBuffer,
+            PIPE_ACCESS_INBOUND | FILE_FLAG_FIRST_PIPE_INSTANCE | FILE_FLAG_OVERLAPPED,
+            PIPE_TYPE_BYTE | PIPE_READMODE_BYTE | PIPE_WAIT | PIPE_REJECT_REMOTE_CLIENTS, 1, 0, kernelBufferSize, 0,
+            nullptr);
+
+        if (!SetFileCompletionNotificationModes(
+                readHandle.get(), FILE_SKIP_COMPLETION_PORT_ON_SUCCESS | FILE_SKIP_SET_EVENT_ON_HANDLE)) {
+            api_failure("SetFileCompletionNotificationModes");
+        }
+
+        SECURITY_ATTRIBUTES inheritSa;
+        inheritSa.nLength              = sizeof(inheritSa);
+        inheritSa.lpSecurityDescriptor = nullptr;
+        inheritSa.bInheritHandle       = TRUE;
+        writeHandle = create_file(pipeNameBuffer, GENERIC_WRITE | FILE_READ_ATTRIBUTES, 0, &inheritSa, OPEN_EXISTING,
+            FILE_FLAG_OVERLAPPED, HANDLE{});
+
+        readIo = tp_io{std::move(readHandle), callback, this, nullptr};
+
+        start();
+    }
+
+    ~output_collecting_pipe() noexcept {
+        if (readIo) {
+            if (running.load()) {
+                if (!CancelIoEx(readIo.get_file(), &overlapped)) {
+                    api_failure("CancelIoEx"); // slams into noexcept
+                }
+            }
+
+            readIo.close();
+        }
+    }
+
+    output_collecting_pipe(const output_collecting_pipe&) = delete;
+    output_collecting_pipe& operator=(const output_collecting_pipe&) = delete;
+
+    void start() {
+        [[maybe_unused]] const auto oldRunning = running.exchange(true);

```

Comment:
(Repeating previous comment) Should this be `bool` instead of `auto`? Ok to leave as-is if you want this.